From 305200b804815611e59dc1367c37c5549c33043e Mon Sep 17 00:00:00 2001
From: Wladimir Leuschner <wladimir.leuschner@qt.io>
Date: Fri, 4 Jul 2025 11:14:50 +0200
Subject: [PATCH 131/553] Windows: Return hittest region for
 CustomizeWindowHint|ExpandedClientAreaHint

When Qt::CustomizeWindowHint in combination with Qt::ExpandedClientAreaHint is set, we are not returning any hittest region for WM_NCHITTEST, resulting in ignoring mouse events. This patch adds another code path to handle WM_NCHITTEST in this case properly. A test whether a Widget was hit is performed with QWSI with the current coordinates and button state of the mouse. In case a widget was hit the event is consumed and NCHITTEST will return HTCLIENT. In case no widget was hit, a check is performed whether the mouse event happened in the titlebar area and if so, the respective hittest region for the titlebar area is returned.

Fixes: QTBUG-137927
Pick-to: 6.9
Change-Id: I4531fec71eaaa20959a8015ea1babd6d36ee9341
Reviewed-by: Oliver Wolff <oliver.wolff@qt.io>
(cherry picked from commit ee2450ccdfac80393b84e773d0d2e0cb6e084019)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../platforms/windows/qwindowswindow.cpp      | 45 +++++++++++++------
 1 file changed, 31 insertions(+), 14 deletions(-)

diff --git a/src/plugins/platforms/windows/qwindowswindow.cpp b/src/plugins/platforms/windows/qwindowswindow.cpp
index 731673b61db..b94c809eda2 100644
--- a/src/plugins/platforms/windows/qwindowswindow.cpp
+++ b/src/plugins/platforms/windows/qwindowswindow.cpp
@@ -3290,6 +3290,7 @@ bool QWindowsWindow::handleNonClientHitTest(const QPoint &globalPos, LRESULT *re
     const QWindow *w = window();
     const QPoint localPos = w->mapFromGlobal(QHighDpi::fromNativePixels(globalPos, w));
     const QRect geom = geometry();
+    static auto oldMouseButtonState = Qt::NoButton;
 
     if (m_data.flags.testFlags(Qt::ExpandedClientAreaHint)) {
         bool isDefaultTitleBar = !w->flags().testFlag(Qt::CustomizeWindowHint);
@@ -3298,17 +3299,18 @@ bool QWindowsWindow::handleNonClientHitTest(const QPoint &globalPos, LRESULT *re
                                                                                                      Qt::WindowMaximizeButtonHint|
                                                                                                      Qt::WindowCloseButtonHint);
         const int border = (IsZoomed(m_data.hwnd) || isFullScreen_sys()) ? 0 : getResizeBorderThickness(savedDpi());
+        const int titleBarHeight = getTitleBarHeight_sys(savedDpi());
+        const int titleButtonWidth = titleBarHeight * 1.5;
+        const bool mouseButtonsSwapped = GetSystemMetrics(SM_SWAPBUTTON);
+        auto mouseButtons = Qt::NoButton;
+        if (mouseButtonsSwapped)
+            mouseButtons = GetAsyncKeyState(VK_LBUTTON) != 0 ? Qt::RightButton : (GetAsyncKeyState(VK_RBUTTON) ? Qt::LeftButton : Qt::NoButton);
+        else
+            mouseButtons = GetAsyncKeyState(VK_LBUTTON) != 0 ? Qt::LeftButton : (GetAsyncKeyState(VK_RBUTTON) ? Qt::RightButton : Qt::NoButton);
+
+        *result = HTCLIENT;
         if (isCustomized || isDefaultTitleBar) {
-            *result = HTCLIENT;
-            const int titleBarHeight = getTitleBarHeight_sys(savedDpi());
-            const int titleButtonWidth = titleBarHeight * 1.5;
             int buttons = 1;
-            const bool mouseButtonsSwapped = GetSystemMetrics(SM_SWAPBUTTON);
-            auto mouseButtons = Qt::NoButton;
-            if (mouseButtonsSwapped)
-                mouseButtons = GetAsyncKeyState(VK_LBUTTON) != 0 ? Qt::RightButton : (GetAsyncKeyState(VK_RBUTTON) ? Qt::LeftButton : Qt::NoButton);
-            else
-                mouseButtons = GetAsyncKeyState(VK_LBUTTON) != 0 ? Qt::LeftButton : (GetAsyncKeyState(VK_RBUTTON) ? Qt::RightButton : Qt::NoButton);
 
             if (globalPos.y() < geom.top() + titleBarHeight) {
                 if (m_data.flags.testFlags(Qt::WindowCloseButtonHint) || isDefaultTitleBar) {
@@ -3336,17 +3338,32 @@ bool QWindowsWindow::handleNonClientHitTest(const QPoint &globalPos, LRESULT *re
                 } if ((isCustomized || isDefaultTitleBar) &&
                       *result == HTCLIENT){
                     QWindow* wnd = window();
-                    if (mouseButtons != Qt::NoButton) {
-                        QMouseEvent event(QEvent::MouseButtonPress, localPos, globalPos, mouseButtons, mouseButtons, Qt::NoModifier);
-                        QGuiApplication::sendEvent(wnd, &event);
-                        if (!event.isAccepted() && mouseButtons == Qt::RightButton)
+                    if (mouseButtons != oldMouseButtonState) {
+                        auto mouseEventType = mouseButtons == Qt::NoButton ? QEvent::MouseButtonRelease : QEvent::MouseButtonPress;
+                        auto mouseEventButtons = mouseEventType == QEvent::MouseButtonPress ? mouseButtons : oldMouseButtonState;
+                        bool accepted = QWindowSystemInterface::handleMouseEvent<QWindowSystemInterface::SynchronousDelivery>(wnd, QHighDpi::toNativeLocalPosition(localPos, w), globalPos, mouseEventButtons, mouseEventButtons, mouseEventType);
+                        if (!accepted && mouseButtons == Qt::RightButton)
                             *result = HTSYSMENU;
-                        else if (!event.isAccepted())
+                        else if (!accepted && globalPos.y() < geom.top() + titleBarHeight)
                             *result = HTCAPTION;
                     }
                 }
             }
+        } else if (w->flags().testFlag(Qt::CustomizeWindowHint)) {
+
+            QWindow* wnd = window();
+            if (mouseButtons != oldMouseButtonState) {
+                auto mouseEventType = mouseButtons == Qt::NoButton ? QEvent::MouseButtonRelease : QEvent::MouseButtonPress;
+                auto mouseEventButtons = mouseEventType == QEvent::MouseButtonPress ? mouseButtons : oldMouseButtonState;
+                bool accepted = QWindowSystemInterface::handleMouseEvent<QWindowSystemInterface::SynchronousDelivery>(wnd, QHighDpi::toNativeLocalPosition(localPos, w), globalPos, mouseEventButtons, mouseEventButtons, mouseEventType);
+                if (!accepted && mouseButtons == Qt::RightButton)
+                    *result = HTSYSMENU;
+                else if (!accepted && globalPos.y() < geom.top() + titleBarHeight)
+                    *result = HTCAPTION;
+            }
         }
+        oldMouseButtonState = mouseButtons;
+
         if (border != 0) {
             const bool left   = (globalPos.x() >= geom.left()) && (globalPos.x() < geom.left() + border);
             const bool right  = (globalPos.x() >  geom.right() - border) && (globalPos.x() <= geom.right());
-- 
2.51.2

